dbms/src/Functions/FunctionsGeo.cpp

```diff
@@ -246,6 +251,181 @@ class FunctionPointInPolygon : public IFunction
+
+        out_offsets.resize(size);
+        out_vec.resize(size * (GEOHASH_MAX_TEXT_LENGTH + 1));
+
+        char * begin = reinterpret_cast<char *>(out_vec.data());
+        char * pos = begin;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            const Float64 latitude = lat_column->getElement(i);
+            const Float64 longitude = lon_column->getElement(i);
+            const Int64 precision = precision_column->getElement(i);
+
+            const size_t encoded_size = GeoUtils::geohashEncode(latitude, longitude, precision, pos);
+
+            pos += encoded_size;
+            *pos = '\0';
+            out_offsets[i] = ++pos - begin;
+        }
+        out_vec.resize(pos - begin);
+
+        if (!out_offsets.empty() && out_offsets.back() != out_vec.size())
+            throw Exception("Column size mismatch (internal logical error)", ErrorCodes::LOGICAL_ERROR);
+
+        result = std::move(col_str);
+
+        return true;
+
+    }
+
+    template <typename LatType, typename LonType>
+    bool tryExecute(const IColumn * lat_column, const IColumn * lon_column, const IColumn * precision_col, ColumnPtr & result)
+    {
+        const ColumnVector<LatType> * latitude = checkAndGetColumnEvenIfConst<ColumnVector<LatType>>(lat_column);
+        const ColumnVector<LonType> * longitude = checkAndGetColumnEvenIfConst<ColumnVector<LonType>>(lon_column);
+        if (!latitude || !longitude)
+            return false;
+
+        return tryExecute2<UInt8>(latitude, longitude, precision_col, result) ||
+               tryExecute2<UInt16>(latitude, longitude, precision_col, result) ||
+               tryExecute2<UInt32>(latitude, longitude, precision_col, result) ||
+               tryExecute2<UInt64>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int8>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int16>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int32>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int64>(latitude, longitude, precision_col, result);
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
+    {
+        const IColumn * latitude = block.getByPosition(arguments[0]).column.get();
+        const IColumn * longitude = block.getByPosition(arguments[1]).column.get();
+        const IColumn * precision = block.getByPosition(arguments[2]).column.get();
+        ColumnPtr & res_column = block.getByPosition(result).column;
+
+        if (tryExecute<Float32, Float32>(latitude, longitude, precision, res_column) ||
+            tryExecute<Float64, Float32>(latitude, longitude, precision, res_column) ||
+            tryExecute<Float32, Float64>(latitude, longitude, precision, res_column) ||
+            tryExecute<Float64, Float64>(latitude, longitude, precision, res_column))
+            return;
```

Comment:
There is an easier way. Try at first check `block.getByPosition(arguments[0]).type` and then choose one of template implementation, and then check the second type and again select one of template implementation, and so on. Like this:

```
    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
    {
        const auto & latitude = block.getByPosition(arguments[0]);
        const auto &longitude = block.getByPosition(arguments[1]).column.get();
        const auto & precision = block.getByPosition(arguments[2]).column.get();
        block.getByPosition(result).column = dispatchForFirstColumn(latitude, longitude, precision);
    }

    ColumnPtr dispatchForFirstColumn(const ColumnWithTypeAndName & latitude, const ColumnWithTypeAndName & longitude, const ColumnWithTypeAndName & precision)
    {
        WhichDataType which_data_type(latitude.type);
        if (which_data.isFloat32())
            return dispatchForSecondColumn(checkAndGetColumn<ColumnFloat32>(latitude.column.get()), longitude, precision);
        if (which_data.isFloat64())
            return dispatchForSecondColumn(checkAndGetColumn<ColumnFloat64>(latitude.column.get()), longitude, precision);
        throw ILLEGAL_TYPE_OF_ARGUMENT
    }

    template <typename LatitudeColumnType>
    ColumnPtr dispatchForSecondColumn(const LatitudeColumnType & latitude, const ColumnWithTypeAndName & longitude, const ColumnWithTypeAndName & precision)
    {
        WhichDataType which_data_type(longitude.type);
        if (which_data.isFloat32())
            return dispatchForThirdColumn(latitude, checkAndGetColumn<ColumnFloat32>(longitude.column.get()), precision);
        if (which_data.isFloat64())
            return dispatchForThirdColumn(latitude, checkAndGetColumn<ColumnFloat64>(longitude.column.get()), precision);
        throw ILLEGAL_TYPE_OF_ARGUMENT
    }

    template <typename LatitudeColumnType, typename LongitudeColumnType>
    ColumnPtr dispatchForThirdColumn(const LatitudeColumnType & latitude, const LongitudeColumnType & longitude, const ColumnWithTypeAndName & precision)
    {
        WhichDataType which_data_type(precision.type);
        if (which_data.isInt32())
            return execute(latitude, longitude, checkAndGetColumn<ColumnInt32>(precision.column.get()));
        if (which_data.isUInt64())
            return execute(latitude, longitude, checkAndGetColumn<ColumnUInt64>(precision.column.get()));
        throw ILLEGAL_TYPE_OF_ARGUMENT
    }

    template <typename LatitudeColumnType, typename LongitudeColumnType, typename PrecisionColumnType>
    ColumnPtr execute(const LatitudeColumnType & latitude, const LongitudeColumnType & longitude, const PrecisionColumnType & precision)
    {
        ...
    }
```

Reply:
> There is an easier way: check `block.getByPosition(arguments[0]).type` and then choose one of template implementations, and then check the second type and again choose one of template implementations. Like this:
> 
> ```
>     void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
>     {
>         const auto & latitude = block.getByPosition(arguments[0]);
>         const auto &longitude = block.getByPosition(arguments[1]).column.get();
>         const auto & precision = block.getByPosition(arguments[2]).column.get();
>         block.getByPosition(result).column = dispatchForFirstColumn(latitude, longitude, precision);
>     }
> 
>     ColumnPtr dispatchForFirstColumn(const ColumnWithTypeAndName & latitude, const ColumnWithTypeAndName & longitude, const ColumnWithTypeAndName & precision)
>     {
>         WhichDataType which_data_type(latitude.type);
>         if (which_data.isFloat32())
>             return dispatchForSecondColumn(checkAndGetColumn<ColumnFloat32>(latitude.column.get()), longitude, precision);
>         if (which_data.isFloat64())
>             return dispatchForSecondColumn(checkAndGetColumn<ColumnFloat64>(latitude.column.get()), longitude, precision);
>         throw ILLEGAL_TYPE_OF_ARGUMENT
>     }
> 
>     template <typename LatitudeColumnType>
>     ColumnPtr dispatchForSecondColumn(const LatitudeColumnType & latitude, const ColumnWithTypeAndName & longitude, const ColumnWithTypeAndName & precision)
>     {
>         WhichDataType which_data_type(longitude.type);
>         if (which_data.isFloat32())
>             return dispatchForThirdColumn(latitude, checkAndGetColumn<ColumnFloat32>(longitude.column.get()), precision);
>         if (which_data.isFloat64())
>             return dispatchForThirdColumn(latitude, checkAndGetColumn<ColumnFloat64>(longitude.column.get()), precision);
>         throw ILLEGAL_TYPE_OF_ARGUMENT
>     }
> 
>     template <typename LatitudeColumnType, typename LongitudeColumnType>
>     ColumnPtr dispatchForThirdColumn(const LatitudeColumnType & latitude, const LongitudeColumnType & longitude, const ColumnWithTypeAndName & precision)
>     {
>         ...
>     }
> ```

Thank you for the tip, I was trying to mimic the rest of the project regarding function type checks.

I do recognize that proposed approach allows to identify which argument exactly is of wrong type, but isn't it the job of `getReturnTypeImpl()` method?

Also I don't see how having 4 bigger functions is easier that having 3 smaller functions.

Reply:
Thanks for the tip, but I was trying to mimic the rest of the project regarding function type checks.

I do recognize that proposed approach allows to identify which argument exactly is of wrong type, but isn't it the job of `getReturnTypeImpl()` method?

Also I don't see how having 4 bigger functions is easier that having 3 smaller functions.

Reply:
No, my comment wasn't about searching wrong argument.
I just meant your code calls multiple implementations and each of them tries to cast columns.
It seems it would be better to check types first and then call one corresponding implementation.

Reply:
well, this part was significantly simplified by making precision always-const parameter.